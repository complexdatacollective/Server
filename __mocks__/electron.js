/* eslint-env jest */

// const electron = jest.genMockFromModule('electron');

const remote = {
  getGlobal: jest.fn(),
};

const ipcRenderer = {
  on: jest.fn(),
};

exports.remote = remote;
exports.ipcRenderer = ipcRenderer;
