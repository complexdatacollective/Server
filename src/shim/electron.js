/* eslint-disable no-console */
/* istanbul ignore next */
(function devBrowserShim() {
  const on = (channel, cb) => { console.log(channel, cb); };
  const once = (channel, cb) => { console.log(channel, cb); };
  const send = (message) => { console.log(message); };

  module.exports = {
    ipcRenderer: {
      send,
      on,
      once,
    },
  };
}());
