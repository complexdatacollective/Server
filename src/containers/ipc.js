import React, { Component } from 'react';
import { ipcRenderer } from 'electron';
import changeCase from 'change-case';

export default (property) => {
  const PROPERTY = changeCase.constantCase(property);

  return (WrappedComponent) => {
    const requestUpdate = () => {
      ipcRenderer.send(`REQUEST_${PROPERTY}`);
    };

    return class IPCComponent extends Component {
      constructor() {
        super();

        this.state = {
          [property]: null,
        };

        this.listen();
        requestUpdate();
      }

      listen() {
        ipcRenderer.on(`${PROPERTY}`, (_, data) => {
          this.setState({
            [property]: data,
          });
        });
      }

      render() {
        return <WrappedComponent {...this.state} {...this.props} />;
      }
    };
  };
};
