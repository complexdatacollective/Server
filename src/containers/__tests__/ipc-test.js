/* eslint-env jest */

import React from 'react';
import { mount, shallow } from 'enzyme';
import { ipcRenderer, remote } from 'electron';
import ipc from '../ipc';

jest.mock('electron');

const globalProperty = 'server';

const MockComponent = () => (
  <div>Intentionally left empty</div>
);

function setup(property = globalProperty) {
  const serverGlobalConnector = ipc(property);
  return {
    component: serverGlobalConnector(MockComponent),
  };
}

describe('ipc HOC', () => {
  beforeEach(() => {
    ipcRenderer.on.mockReset();
  });

  it('It should render', () => {
    const Component = setup().component;
    const subject = shallow(<Component />);

    expect(subject).toMatchSnapshot();
  });

  it('It should return a HOC Wrapped Component', () => {
    remote.getGlobal
      .mockReturnValueOnce({ foo: 'bar' })
      .mockReturnValueOnce({ baz: 'buzz' });

    const Component = setup().component;

    const subject = mount((
      <Component foo="bar" />
    )).find('MockComponent');

    expect(subject.prop(globalProperty)).toEqual({ foo: 'bar' });

    const globalUpdatedCallback = ipcRenderer.on.mock.calls[0][1];
    globalUpdatedCallback();

    expect(subject.prop(globalProperty)).toEqual({ baz: 'buzz' });
  });
});
