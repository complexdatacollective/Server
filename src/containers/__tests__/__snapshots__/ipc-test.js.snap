exports[`ipc HOC It should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <MockComponent
    server={undefined} />,
  "nodes": Array [
    <MockComponent
      server={undefined} />,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <IPCComponent />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": IPCComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "server": undefined,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <MockComponent
          server={undefined} />,
        "_debugID": 2,
        "_renderedOutput": <MockComponent
          server={undefined} />,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <IPCComponent />,
}
`;
